### ISSUE TITLE
Stage selected lines

### ISSUE THREAD

**Issue by infabo** [sublimehq/sublime_merge#967]
**Problem description**

Suggested "hunks" are too large. There are changes on 3 different sections - but only one big hunk.

![image](https://user-images.githubusercontent.com/3999104/97676797-79254d80-1a91-11eb-9688-04a8bcbb3e1b.png)

**Preferred solution**

I would like to stage individual lines. Either by visually selecting the lines I want to stage, or by right-click on a specific line.

For reference: GitExtensions for Windows has this feature ever since:

![image](https://user-images.githubusercontent.com/3999104/97677439-8858cb00-1a92-11eb-8181-92a7f4cd5832.png)

**Comment by stinos** [comment]
Select a line, then the 'Stage Line' button appears

**Comment by dpjohnst** [comment]
Hi @infabo,

Thanks for sharing your feedback!

As @stinos mentioned, this is already possible using the steps provided.

Let me know if you have any difficulties in staging lines, and I can definitely assist further.

Kind regards,
\- Dylan

**Comment by infabo** [comment]
OMG! That's super bad UX.

**Comment by infabo** [comment]
I would have never ever expected, that an existing button changes its label and its behaviour. I have not experienced this in any desktop-software in my whole life.

**Comment by stinos** [comment]
I'd wager that's merely an opinion. The alternatives aren't exactly that much better?  Doubling the number of buttons seems like a no-go, hiding it in a right-click menu is also not exactly supper good. Also not that the current is still what practically is the fastest: select with mouse, stage/discard with keyboard. That's faster than selecting with mouse and then going into a menu.

**Comment by infabo** [comment]
Considering my 27" 1440p screen, it's quite unlikely when you select lines on the LEFT side - some label change on the right side of the screen. Thats out of visual scope IMHO.

![image](https://user-images.githubusercontent.com/3999104/97704693-2f4d5f00-1ab3-11eb-9a26-ba68106ded19.png)

I would suggest to add "stage linkes" to context-menu when you right-click on the line-selection.

**Comment by infabo** [comment]
> hiding it in a right-click menu is also not exactly supper good

It's way better. Thats what I think - at least Windows-people - learned since Windows 95. Right-Click reveals options to your selection. That's trained for 20 years.

**Comment by infabo** [comment]
>  Also note that the current is still what practically is the fastest: select with mouse, stage/discard with keyboard. 

I agree with that. Now I learned `Enter` stages the hunk/line.

**Comment by stinos** [comment]
>  on the LEFT side - some label change on the right side of the screen. Thats out of visual scope IMHO

That's fair, but that's just the initial discovery which unless you very easily forget things has to be done once only, like for any other user interface action. Also it's a bit more nuanced: when working with a selection you do have to look both at the left and right sides of the diff anyway to verify what's selected, otherwise you might be staging/discarding the wrong thing (in case of a simple add/remove like what you have there it doesn't matter but add/remove and changes make this non-trivial to guess). In other words visual scope is both sides of the diff. And you can also select on the right side btw. Another reason this isn't that bad: once you select something it makes sense you want to do something with that selection, so the context changes which results in different commands.

> I would suggest to add "stage linkes" to context-menu when you right-click on the line-selection

 Right-click menu doesn't have Stage hunk either so that should be added as well then.

**Comment by infabo** [comment]
> Also it's a bit more nuanced: when working with a selection you do have to look both at the left and right sides of the diff anyway to verify what's selected,

Yep. Maybe my example with the deletion of lines wasn't the most common one.

**Comment by infabo** [comment]
> Right-click menu doesn't have Stage hunk either so that should be added as well then.

+1

**Comment by infabo** [comment]
> That's fair, but that's just the initial discovery which unless you very easily forget things has to be done once only

I found it hard to discover.

Considering that I bought my Sublime Merge license 09.11.2018 - and evaluated it several weeks before acquiring the license...

...it took me 2 years to discover this feature?

Either I am super-dumb or as already said: bad UX.

**Comment by stinos** [comment]
> ...it took me 2 years to discover this feature?

I find it hard to believe that's really what happened. Don't get me wrong, I'm not going to say it's extremely easy to discover, it isn't, but your logic seems a bit off: you see there is only one way to stage/discard lines, and that is with those buttons (and the Enter/Backspace keys bound to it). So instead of taking it you 2 years to discover the right-click menu is missing, wouldn't it rather be the case you didn't need/want to stage selected lines only and just didn't search for how to do it? I.e. I can't believe you've been selecting lines for 2 years and kept on wondering 'how do I do something with it' and not once see the button? Or even if that was the case, you could have searched for it (first google match for 'sublime merge stage line' is spot on)? Again: that only proves it's not the easiest to discover if you have to search for it, but that's not the same as 'took 2 years to discover'

**Comment by stinos** [comment]
What would help though, is slightly chaning the button color upon making a selection, that's much harder to miss than changing text.

**Comment by stinos** [comment]
> A simple solution for that might be to display a checkbox on each side for each modified line. Clicking the checkbox or the line number could toggle the selection state of the checkbox. Toggling the checkbox might either stage/unstage the line directly, or simply mark it for staging.

That sounds a bit messy, checkboxes for all lines. Currently the workaround for your issue is simple to do, but not simple to discover and sort of a side-effect. In the preferences set 'diff view style' to 'auto'. Then following the example you show change the size of the diff view horizontally, by making the main SM window less wide for instance, and then SM automatically switches to the inline diff view instead of the side-by-side diff view. In inline diff view you can stage/unstage individual lines.

tldr; what SM needs is a simple way to toggle diff view styles

**Comment by infabo** [comment]
> > ...it took me 2 years to discover this feature?
> 
> I find it hard to believe that's really what happened.

Seems like some other dude had that problem too.

![image](https://user-images.githubusercontent.com/3999104/117659572-660bdc80-b19c-11eb-8a02-e8e24cc3b7ff.png)

**Comment by infabo** [comment]
> > ...it took me 2 years to discover this feature?
> 
> I find it hard to believe that's really what happened. Don't get me wrong, I'm not going to say it's extremely easy to discover, it isn't, but your logic seems a bit off: you see there is only one way to stage/discard lines, and that is with those buttons (and the Enter/Backspace keys bound to it). So instead of taking it you 2 years to discover the right-click menu is missing, wouldn't it rather be the case you didn't need/want to stage selected lines only and just didn't search for how to do it? I.e. I can't believe you've been selecting lines for 2 years and kept on wondering 'how do I do something with it' and not once see the button? Or even if that was the case, you could have searched for it (first google match for 'sublime merge stage line' is spot on)? Again: that only proves it's not the easiest to discover if you have to search for it, but that's not the same as 'took 2 years to discover'

First, I do read the changelogs when they pop-up after an update. I did not recognize such a new feature there. Maybe stage-lines was already there in 2018 - maybe - but I did not find it. So instead of trying over and over again I just decided to use GitExtensions to stage individual line - **when** I was in need of this feature. Maybe happened once a day or once in a week. As I still have difficultied to "read" Sublime Merge's git-branch-graph - I still use other tools like GitExtensions to visually display the branch-graph in repositories with many feature-branches. So short: I did not find the feature back then, I am sorry for not googling it (but maybe back in 2018 I maybe haven't found anything on the topic though) and I am happily using "stage lines" since you pointed me to it. So all good. ;)

**Comment by doerwalter** [comment]
> > A simple solution for that might be to display a checkbox on each side for each modified line. Clicking the checkbox or the line number could toggle the selection state of the checkbox. Toggling the checkbox might either stage/unstage the line directly, or simply mark it for staging. ... The buttons "Stage" and "Stage hunk" get changed into "Select all" and "Select hunk", they don't stage the lines directly, but simply mark them
> 
> That sounds a bit messy, checkboxes for all lines, and/or also a lot of different steps to get the job done.

Only modified lines would need a checkbox.

> Currently the workaround for your issue is simple to do, but not simple to discover and sort of a side-effect. In the preferences set 'diff view style' to 'auto'. Then following the example you show change the size of the diff view horizontally, by making the main SM window less wide for instance, and then SM automatically switches to the inline diff view instead of the side-by-side diff view. In inline diff view you can stage/unstage individual lines.

Indeed a useful trick! And using "Text Diff Style" to "inline" would probably be even simpler.

> tldr; what SM needs is a simple way to toggle diff view styles

Another approach would be a new configuration option "Don't synchronize selection when staging". I.e. then I could select lines on both sides independently without having Sublime Merge autoselect lines on the other side.

**Comment by stinos** [comment]
That would also be interesting! But if it is just a configuration option (as opposed to being able to quickly toggle it, or e.g. buttons like `Stage Left` and `Stage Right`, just thinking out loud here) then it's either one or the other which would reduce usability for those who want to keep the current `Stage Lines` functionality which is definitely also convenient.

**Comment by doerwalter** [comment]
> That would also be interesting! But if it is just a configuration option (as opposed to being able to quickly toggle it, or e.g. buttons like `Stage Left` and `Stage Right`, just thinking out loud here) then it's either one or the other which would reduce usability for those who want to keep the current `Stage Lines` functionality which is definitely also convenient.

True, even if sometimes you want to stage something from the left or the right side only, in most cases you want to stages both sides simultaneously. An always having to reconfigure Sublime Merge for that is inconvenient.

So mayby switching between "select lines from both sides" and "select lines from this side only" can be toggled via a qualifier key (ctrl, shift, alt etc.).
